core/shared/src/main/scala/org/typelevel/idna4s/core/CodePoint.scala

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2022 Typelevel
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package org.typelevel.idna4s.core
+
+import cats.kernel._
+import cats.Show
+import cats.syntax.all._
+
+/**
+ * A newtype for representing a Unicode code point.
+ *
+ * Unicode code points are a integral values from [0, 0x10FFF].
+ *
+ * This type is not (and should not) generally in critical path performance sensitive code. It
+ * is mostly useful for generating error information, testing, and for some performance
+ * insensitive methods which operate only a single code point. In particular for errors messages
+ * where we generally should not render arbitrary code points as they may influence the output
+ * text in ways we do not anticipate, such as bidi scripts or where the code point in isolation
+ * is not able to be rendered at all.
+ *
+ * @see
+ *   [[https://www.unicode.org/reports/tr36/ Unicode Security Considerations]]
+ */
+sealed abstract class CodePoint extends Serializable {
+
+  /**
+   * The integral value of the code point.
+   */
+  def value: Int
+
+  // final
+
+  /**
+   * The number of UTF-16 bytes needed to represent this code point. UTF-16 encodes code points
+   * >= 0x10000 as surrogate pairs, requiring two bytes to represent.
+   */
+  final def charCount: Int =
+    if (value >= 0x10000) 2 else 1
+
+  /**
+   * Convert this code point to a list of utf-16 bytes (char values), either size 1 or 2
+   * depending on if this code point is represents a surrogate pair in UTF-16.
+   */
+  final def toChars: List[Char] =
+    Character.toChars(value).toList
+
+  final override def toString: String =
```

Comment:
Maybe worth adding a javadoc that for security and other reasons (?), this doesn't directly render the codepoint itself, but only metadata about it.

Reply:
:+1: I'll add a doc. It's both security and readability.